specify/specify7
# Query Builder Distinct Option Returns Links to Record IDs

Fixes #4607 

After talking with Andy, he thought that checking the distinct checkbox would work to fix #1450, if it provided links to each record. 

The motive of this solution is to avoid the user seeing multiple identical rows next to each other in the query results.  In order to avoid duplicate records like in #1450, this solution takes all of the display fields checked by the user and adds them to a GROUP BY statement, thus avoiding duplicates.  This is done instead of using DISTINCT so that a list of the record IDs can be returned.

This solution's idea came from https://github.com/specify/specify7/issues/1464#issuecomment-1286985113
> When running a distinct query, can we mimic a distinct query using "GROUP BY" instead of "SELECT DISTINCT" and include the id in the query as a `CONCAT(id, ", ")` (which would be parsed on the front-end into separate records)?

I think this solution is a good idea, and implementing a front-end solution to handle a `GROUP_CONCAT(id, ", ")` field in the record set could be nice.  So the front-end code needs to be able to handle multiple record IDs.  When showing the results, each row will need to provide links to all the associated record IDs.

The difference from this branch, is that the issue-1450-2 branch moves the record ID from the GROUP_CONCAT statement to the the group by statement.  The current design could be modified, but we need to be mindful of any other side effects in changing the original design intentions https://github.com/specify/specify7/blob/67d73f4663cfef67356e95487da61cae1ae43db9/specifyweb/stored_queries/execution.py#L577

___
***Edit:
When displaying distinct query results, it will have a column to the record id, so it will look the same as non-distinct queries. The difference if that we're going to have a dialog box, when clicking on the record form link, that will allow you to pick one of the records associated with that query result row. @realVinayak had the idea to base it on the pop-up dialog from [agent view](https://sp7demofish.specifycloud.org/specify/view/agent/66/), which appears when clicking on the 'Delete' button.
___

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

Use the query builder to make queries like those in #1450 and make sure that there are no duplicate records in the results.  Also try to make sure that the file created by export csv looks good.

Any query known to create duplicates would be a good test case for this PR.  Try running as many queries as you can using the distinct mode.
